### DISCUSSION TITLE
Custom Item Model Transformation and Depth

### DISCUSSION THREAD

**Discussion by Bawnorton** [FabricMC/fabric-api#3125] [Mod Dev Support]
### Description

I am trying to generate custom item models for armour such that I can render the trims on top of them for any colour, thus, I am using a custom model for "generating" the texture via `FabricBakedModel#emitItemQuads` However, the model appears different to what I expect when rendered in the world / in the player's hands as the vanilla model:

Actual:
Video: https://github.com/FabricMC/fabric/assets/18416784/474e20f7-8a1a-44c5-86e8-8406feeef61c
<img width="1469" alt="Screenshot 2023-06-15 at 2 15 17 PM" src="https://github.com/FabricMC/fabric/assets/18416784/83303707-0f08-4b37-b786-59862bdf2ca0">

Expected: 
Video: https://github.com/FabricMC/fabric/assets/18416784/5978f04a-6883-4663-a7cc-11d58a7a6e09
<img width="1469" alt="Screenshot 2023-06-15 at 2 15 27 PM" src="https://github.com/FabricMC/fabric/assets/18416784/e9040b94-610d-46a7-bbf3-a12bf7f05735">

<details>
  <summary>== Model Class ==</summary>

```java
public class ColouredArmorTrimModel implements UnbakedModel, BakedModel, FabricBakedModel {
    public static JsonUnbakedModel ITEM_GENERATED = null;

    private final SpriteIdentifier SPRITE_IDENTIFIER;

    private ArmorItem ARMOUR_ITEM;
    private Sprite ARMOUR_SPRITE;
    private Sprite OVERLAY_SPRITE;
    private Sprite TRIM_SPRITE;

    private Mesh mesh;

    private ColouredArmorTrimModel(Identifier armourId, Identifier resourceId) {
        ARMOUR_ITEM = (ArmorItem) Registries.ITEM.get(armourId);
        SPRITE_IDENTIFIER = new SpriteIdentifier(SpriteAtlasTexture.BLOCK_ATLAS_TEXTURE, resourceId);
    }

    public static ColouredArmorTrimModel create(Identifier armourId, Identifier resourceId) {
        return new ColouredArmorTrimModel(armourId, resourceId);
    }

    @Override
    public boolean isVanillaAdapter() {
        return false;
    }

    @Override
    public void emitBlockQuads(BlockRenderView blockView, BlockState state, BlockPos pos, Supplier<Random> randomSupplier, RenderContext context) {
        context.meshConsumer().accept(mesh);
    }

    @Override
    public void emitItemQuads(ItemStack stack, Supplier<Random> randomSupplier, RenderContext context) {
        QuadEmitter emitter = context.getEmitter();
        Item item = stack.getItem();
        if (!(item instanceof ArmorItem armourItem) || !armourItem.equals(ARMOUR_ITEM)) {
            throw new AssertionError("ColouredArmorTrimModel received an ItemStack that was not an ArmorItem of the correct type. Got " + item.getName().getString() + " but expected " + ARMOUR_ITEM.getName().getString() + ".");
        }

        int colour = -1;
        if(armourItem instanceof DyeableArmorItem dyeableArmorItem) {
            colour = dyeableArmorItem.getColor(stack) - 0xFFFFFF - 1;
        }

        emitter.square(Direction.SOUTH, 0, 0, 1, 1, 0);
        emitter.spriteBake(ARMOUR_SPRITE, MutableQuadView.BAKE_LOCK_UV);
        emitter.color(colour, colour, colour, colour);
        emitter.emit();

        context.meshConsumer().accept(mesh);
    }

    @Override
    public List<BakedQuad> getQuads(@Nullable BlockState state, @Nullable Direction face, Random random) {
        return Collections.emptyList();
    }

    @Override
    public boolean useAmbientOcclusion() {
        return false;
    }

    @Override
    public boolean hasDepth() {
        return true;
    }

    @Override
    public boolean isSideLit() {
        return false;
    }

    @Override
    public boolean isBuiltin() {
        return false;
    }

    @Override
    public Sprite getParticleSprite() {
        return ARMOUR_SPRITE;
    }

    @Override
    public ModelTransformation getTransformation() {
        return ITEM_GENERATED.getTransformations();
    }

    @Override
    public ModelOverrideList getOverrides() {
        return ModelOverrideList.EMPTY;
    }

    @Override
    public Collection<Identifier> getModelDependencies() {
        return Collections.emptyList();
    }

    @Override
    public void setParents(Function<Identifier, UnbakedModel> modelLoader) {
        ITEM_GENERATED = (JsonUnbakedModel) modelLoader.apply(new Identifier("minecraft", "item/generated"));
    }

    @Nullable
    @Override
    public BakedModel bake(Baker baker, Function<SpriteIdentifier, Sprite> textureGetter, ModelBakeSettings rotationContainer, Identifier modelId) {
        ARMOUR_SPRITE = textureGetter.apply(SPRITE_IDENTIFIER);

        Renderer renderer = RendererAccess.INSTANCE.getRenderer();
        if(renderer == null) return this;

        MeshBuilder builder = renderer.meshBuilder();
        QuadEmitter emitter = builder.getEmitter();
        emitter.emit();
        mesh = builder.build();
        return this;
    }
}
```
</details>

Any help would be appreciated.

**Comment by Technici4n** [comment]
You'll need to override `ModelTransformation getTransformation()` to return the transformation you want.

**Reply by Bawnorton** [reply]
I am, I'm using:
```
    @Override
    public ModelTransformation getTransformation() {
        return ITEM_GENERATED.getTransformations();
    }
```

where ITEM_GENERATED is defined as so:
```
    @Override
    public void setParents(Function<Identifier, UnbakedModel> modelLoader) {
        ITEM_GENERATED = (JsonUnbakedModel) modelLoader.apply(new Identifier("minecraft", "item/generated"));
    }
```

which is the same parent the normal armour models use afaik

**Reply by Technici4n** [reply]
Ah, in that case I have no idea

**Comment by BenjaminSipe** [comment]
For anyone looking at this later. 
The answer is that it is actually very complicated and not worth implementing by hand.

Luckily there is a ItemModelGenerator class, that can be paired with a little functionality of the ModelLoader class to get the desired result.

I'm too lazy to try and fix the code above, but I'll give some code to point people in the right direction.
1. For starters, use the when you get the SPRITE_IDENTIFIER, you should also go ahead and load a JsonUnbakedModel for the Sprite as well. . . 
ATM, I'm doing this via a hardcoded json string. . . but ModelLoader also has a private function that loads it from the file directly. 

2. Just use ItemModelGenerator's implementation. . . 

```
// near the top
ItemModelGenerator ITEM_MODEL_GENERATOR = new ItemModelGenerator();

// in your emitItemQuads function. 
JsonUnbakedModel model = deserialize( shortFormJson );
BakedModel bakedModel = ITEM_MODEL_GENERATOR.create(this.textureGetter, model).bake(this, model, textureGetter, this.modelBakeSettings, false);
bakedModel.emitItemQuads( itemStack, supplier, renderContext );
```

For this to work you will also have to have dummy implementations for Baker. . . 

There are also some local variables I'm just assuming you'll figured out. 

Note: from what I can tell, fabric ( and several mods I've torn apart ) all make sure to implement some kind of cache in this process to avoid re-creating BakedModels for the same items twice. So it's highly suggested you make some kind of cache as well to avoid making these over and over again.

Lastly, answering the specific question here. . . from the sounds of things you want to render the armor, and render a trim sprite over top. . . 
In which case, do what I mention above _along side_ what you are doing at the moment. Make sure to render the outermost item last. . . But it should work.
